pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,484 @@
+		return
+	}
+
+	t.setConnectionInProgress(node)
+	defer t.clearConnectionInProgress(node)
+
+	t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %v is connecting to node %v", t.ownID, node))
+
+	info, found := t.getNodeAddrInfo(node)
+	if !found {
+		t.logger.Log(logging.LevelError, "failed to get node address", "src", t.ownID, "dst", node)
+		return
+	}
+
+	t.host.Peerstore().AddAddrs(info.ID, info.Addrs, PermanentAddrTTL)
+
+	s, err := t.openStream(ctx, info.ID)
+	if err != nil {
+		t.logger.Log(logging.LevelError, "failed to open stream to node", "addr", info, "node", node, "err", err)
+		return
+	}
+
+	t.addOutboundStream(node, s)
+	t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s has connected to node %s", t.ownID.Pb(), node.Pb()))
+}
+
+func (t *Transport) openStream(ctx context.Context, dest peer.ID) (network.Stream, error) {
+	// We need the simplest retry mechanism due to the fact that the underlying libp2p's NewStream function dials once:
+	// https://github.com/libp2p/go-libp2p/blob/7828f3e0797e0a7b7033fa5e8be9b94f57a4c173/p2p/net/swarm/swarm.go#L358
+	t.logger.Log(logging.LevelDebug, "opening stream to peer", "src", t.ownID, "dst", dest)
+
+	// The implementation is based on the openStream() function from the RemoteTracer:
+	// https://github.com/libp2p/go-libp2p-pubsub/blob/cbb7bfc1f182e0b765d2856f6a0ea73e34d93602/tracer.go#L280
+	var s network.Stream
+	var err error
+	for i := 0; i < maxRetries; i++ {
+		sctx, cancel := context.WithTimeout(ctx, maxConnectingTimeout)
+
+		s, err = t.host.NewStream(sctx, dest, ProtocolID)
+		cancel()
+
+		if err == nil {
+			return s, nil
+		}
+
+		if i >= noLoggingErrorAttempts {
+			t.logger.Log(
+				logging.LevelError, fmt.Sprintf("%s failed to open stream to %s: %v", t.ownID, dest, err))
+		} else {
+			t.logger.Log(
+				logging.LevelInfo, fmt.Sprintf("%s failed to open stream to %s: %v", t.ownID, dest, err))
+		}
+
+		delay := time.NewTimer(maxRetryTimeout)
+		select {
+		case <-delay.C:
+			continue
+		case <-ctx.Done():
+			if !delay.Stop() {
+				<-delay.C
+			}
+			return nil, fmt.Errorf("%s opening stream to %s: context closed", t.ownID, dest)
+		}
+	}
+	return nil, fmt.Errorf("%s failed to open stream to %s: %w", t.ownID, dest, err)
+}
+
+func (t *Transport) sendPayload(dest types.NodeID, payload []byte) error {
+	t.nodesLock.RLock()
+	defer t.nodesLock.RUnlock()
+
+	// There are two cases when we get an error:
+	// 1. We don't have the node ID in the nodes table. E.g. we didn't call Connect().
+	// 2. We created an entry for the node but have not opened a connection.
+	// But if we added the node via Connect, then it should open the connection.
+	// It doesn't make sense to open a new one here.
+	stream, err := t.getNodeStreamAndInfoWithoutLock(dest)
+	if err != nil {
+		return err
+	}
+
+	w := bufio.NewWriter(stream)
+	if _, err = w.Write(payload); err != nil {
+		return err
+	}
+	return w.Flush()
```

Comment:
What is the advantage of using `bufio.Writer` here?

Reply:
That is a common [approach](https://github.com/libp2p/go-libp2p/tree/master/examples) for libp2p applications.

Reply:
What would you recommend to use instead of that?

Reply:
Why not writing to the libp2p stream directly?

Reply:
fixed